pkg/backend/httpstate/state.go

```diff
@@ -156,92 +157,76 @@ func (u *cloudUpdate) Complete(status apitype.UpdateStatus) error {
 	return u.backend.client.CompleteUpdate(u.context, u.update, status, token)
 }
 
-// recordEngineEvent will record the event with the Pulumi Service, enabling things like viewing
+// recordEngineEvents will record the events with the Pulumi Service, enabling things like viewing
 // the update logs or drilling into the timeline of an update.
-func (u *cloudUpdate) recordEngineEvent(event engine.Event, sequenceNumber int) error {
+func (u *cloudUpdate) recordEngineEvents(startingSeqNumber int, events []engine.Event) error {
 	contract.Assert(u.tokenSource != nil)
 	token, err := u.tokenSource.GetToken()
 	if err != nil {
 		return err
 	}
 
-	apiEvent, convErr := convertEngineEvent(event)
-	if convErr != nil {
-		return errors.Wrap(convErr, "converting engine event")
-	}
+	var apiEvents apitype.EngineEventBatch
+	for idx, event := range events {
+		apiEvent, convErr := convertEngineEvent(event)
+		if convErr != nil {
+			return errors.Wrap(convErr, "converting engine event")
+		}
 
-	// Each event within an update must have a unique sequence number. Any request to
-	// emit an update with the same sequence number will fail. (Read: use a mutex to
-	// increment if needed.)
-	apiEvent.Sequence = sequenceNumber
-	apiEvent.Timestamp = int(time.Now().Unix())
+		// Each event within an update must have a unique sequence number. Any request to
+		// emit an update with the same sequence number will fail. (Read: the caller needs
+		// to be accurate about this.)
+		apiEvent.Sequence = idx + startingSeqNumber
+		apiEvent.Timestamp = int(time.Now().Unix())
 
-	return u.backend.client.RecordEngineEvent(u.context, u.update, apiEvent, token)
-}
+		apiEvents.Events = append(apiEvents.Events, apiEvent)
+	}
 
-func isDebugDiagEvent(e engine.Event) bool {
-	return e.Type == engine.DiagEvent && (e.Payload.(engine.DiagEventPayload)).Severity == diag.Debug
+	return u.backend.client.RecordEngineEvents(u.context, u.update, apiEvents, token)
 }
 
 // RecordAndDisplayEvents inspects engine events from the given channel, and prints them to the CLI as well as
 // posting them to the Pulumi service.
 func (u *cloudUpdate) RecordAndDisplayEvents(
 	label string, action apitype.UpdateKind, stackRef backend.StackReference, op backend.UpdateOperation,
 	events <-chan engine.Event, done chan<- bool, opts display.Options, isPreview bool) {
+	// We take the channel of engine events and pass them to separate components that will display
+	// them to the console or persist them on the Pulumi Service. Both should terminate as soon as
+	// they see a CancelEvent, and when finished, close the "done" channel.
+	displayEvents := make(chan engine.Event, 100)
```

Comment:
Was there a reason besides symmetry for making the displayEvents channel buffered?

Reply:
Besides symmetry? No. I'll remove it and leave the `displayEvents` channel unbuffered. In theory it could lead to us delaying persisting engine events if the `displayEvents` channel doesn't pulled from. But I assume that there is little-to-no measurable delay in writing to STDOUT.